### PR TITLE
refactor: replace embedded enjoi fork with ajv-backed schema validator

### LINKED ISSUE #5384: Warning in container logs using `/chromium/pdf` REST API
**Describe the bug**
There is waring in container logs using `/chromium/pdf` REST API:
```
WARNING: schema missing a 'type' or '$ref' or 'enum': 
{}
```

**To Reproduce**
```bash
curl --data '{"html": "test"}' -H 'Content-Type: application/json' 'http://localhost:3000/chromium/pdf?token=some_token'
```

**Expected behavior**
No warning in container logs

**Version**
v2.49.0

### PR DESCRIPTION
**Pull request by artiom**
## Summary
- Replace the embedded `enjoi` fork (unmaintained upstream) with `ajv` via a new `schema-validator.ts`. Joi-compatible API surface preserved.
- Drop `joi`, `@hapi/bourne`, `@hapi/hoek`. Add `ajv`. Net −1 direct dep, −4 transitive packages.
- Closes #5384 — ajv treats empty `{}` schemas as "accept any" per spec, silently.
- Coercion parity with joi+enjoi verified: nested string→number/boolean, `launch=anyOf<object,string>`, `additionalProperties: {}`, anyOf declared-order semantics.
- Add `schema-coercion-parity.spec.ts` (9 cases). Bodies are taken from canonical examples on docs.browserless.io.
- `scripts/build-schemas.js` untouched; generated JSON and `swagger.json` byte-identical to `main`.
- Old path stays as an 11-line passive re-export shim for external SDK consumers.

## Test plan
- [x] `npm run build` — exit 0
- [x] `npm run lint` — 0 errors
- [x] `npm test` — 351 passing.
- [x] `static/docs/swagger.json` and `build/routes/**/*.json` byte-identical to `main`
- [x] SDK scaffold + custom route: build emits schemas, runtime accepts valid input + rejects bad input (5/5), zero warnings

🤖 Generated with [Claude Code](https://claude.com/claude-code)

---

<a href="https://app.devin.ai/review/browserless/browserless/pull/5388" target="_blank">
  <picture>
    <source media="(prefers-color-scheme: dark)" srcset="https://static.devin.ai/assets/gh-open-in-devin-review-dark.svg?v=1">
    Open in Devin Review
  </picture>
</a>

## Summary by CodeRabbit

* **Refactor**
  * Validation consolidated to a single JSON‑Schema-based validator for consistent parsing and coercion.

* **Tests**
  * Expanded contract and unit tests for request/query/body validation, coercion edge cases, alternative selection, and prototype‑pollution regressions.

* **Chores**
  * Updated dependency set to support the new validator and removed legacy schema libraries.

* **Bug Fixes**
  * More consistent/stricter validation, clearer error responses, and hardened JSON parsing.

[Review Change Stack](https://app.coderabbit.ai/change-stack/browserless/browserless/pull/5388?utm_source=github_walkthrough&utm_medium=github&utm_campaign=change_stack)

### CODE REVIEW

File: src/shared/utils/schema-validator.ts [hunk 1]
```diff
@@ -0,0 +1,359 @@
...
+): unknown | undefined => {
+  if (expected === 'number' || expected === 'integer') {
+    if (value === '') return undefined;
+    const n = Number(value);
+    if (!Number.isFinite(n)) return undefined;
+    if (expected === 'integer' && !Number.isInteger(n)) return undefined;
+    return n;
+  }
+  if (expected === 'boolean') {
+    if (value === 'true' || value === '') return true;
```
[the remark this answers is not in the document]
Comment (artiom) [reply src/shared/utils/schema-validator.ts]: Fixed in 73f96d38. Removed the `|| value === ""` branch in `coerceStringToType` for boolean (`schema-validator.ts:100-104`). Empty strings are now rejected, matching joi 18 behavior. Added an explicit regression test "rejects empty string for a nested boolean field" in `schema-coercion-parity.spec.ts`.

File: src/shared/utils/schema-validator.ts [hunk 2]
```diff
@@ -0,0 +1,359 @@
...
+        return coerceAgainstSchema(input, arrayAlt, root, depth + 1, refStack);
+      }
+      return input;
+    }
+    if (input && typeof input === 'object') {
+      const objAlt = alts.find((a) => inferExpectedType(a, root) === 'object');
+      if (objAlt) {
+        return coerceAgainstSchema(input, objAlt, root, depth + 1, refStack);
+      }
+      return input;
```
[the remark this answers is not in the document]
Comment (artiom) [reply src/shared/utils/schema-validator.ts:269]: Fixed in 73f96d38. Both the array and object branches now iterate alternatives in declared order and return the first that produces a coercion (`schema-validator.ts:195-211`). Uses `result !== input` as the "did this alt accept" signal; the recursive walker only allocates a new reference when a descendant was actually coerced, so the check is faithful.

File: src/shared/utils/schema-validator.ts [hunk 3]
```diff
@@ -0,0 +1,359 @@
...
+    }
+    return input;
+  }
+
+  const expected = inferExpectedType(schema, root);
+
+  if (typeof input === 'string') {
+    if (!expected || expected === 'string' || expected === 'null') return input;
+    const coerced = coerceStringToType(input, expected);
+    return coerced !== undefined ? coerced : input;
```
[the remark this answers is not in the document]
Comment (artiom) [reply src/shared/utils/schema-validator.ts]: Fixed in 73f96d38 using the suggested patch (`schema-validator.ts:214-222`). After `coerceStringToType` returns the parsed object/array, the code now recurses so nested stringified primitives also coerce. Locked in by the new test "recurses after parsing a JSON-string launch so nested fields coerce too".

File: src/shared/utils/schema-validator.ts [hunk 4]
```diff
@@ -0,0 +1,359 @@
...
+  constructor(jsonSchema: object) {
+    this.validator = getCompiled(jsonSchema);
+    this.schema = jsonSchema as Record<string, unknown>;
+  }
+
+  // _opts is accepted for joi API parity (e.g., { abortEarly }); ajv always returns all errors.
+  // eslint-disable-next-line @typescript-eslint/no-unused-vars
+  validate<T = unknown>(input: unknown, _opts?: ValidateOptions): ValidationResult<T> {
+    const coerced = coerceAgainstSchema(input, this.schema, this.schema, 0, new Set());
+    const valid = this.validator(coerced);
```
[the remark this answers is not in the document]
Comment (artiom) [reply src/shared/utils/schema-validator.ts:383]: Alert dismissed as `won't fix`. `allErrors: true` preserves the prior validator's multi-error reporting (joi `abortEarly: false`); the CWE-400 vector is bounded by the HTTP body size cap (`Config.getMaxPayloadSize()` enforced before validation). Mitigation now documented at `schema-validator.ts:18-21`.

File: src/server.ts [hunk 5]
```diff
@@ -267,10 +267,8 @@ export class HTTPServer extends EventEmitter {
     if (route.querySchema) {
       this.logger.trace(`Validating route query-params with QUERY schema`);
       try {
-        const schema = EnjoiResolver.schema(route.querySchema);
-        const valid = schema.validate(req.queryParams, {
-          abortEarly: false,
-        });
+        const schema = compileSchema(route.querySchema);
+        const valid = schema.validate(req.queryParams);
```
[the remark this answers is not in the document]
Comment (artiom) [reply src/server.ts:271]: Fixed in eddf8a54 using the suggested patch. All three call sites (`server.ts:298` HTTP query, `server.ts:341-342` HTTP body, `server.ts:437` WS query) now assign `valid.value` back to `req.queryParams` / `req.body`. Verified end-to-end with a live SDK test: sending `{name:"X", count:"3"}` to a route with `count: number` now reaches the handler as `3` (number).

File: src/shared/utils/schema-validator.ts [hunk 6]
```diff
@@ -0,0 +1,351 @@
...
+        if (result !== input) return result;
+      }
+      return input;
+    }
+    if (input && typeof input === 'object') {
+      for (const alt of alts) {
+        if (inferExpectedType(alt, root) !== 'object') continue;
+        const result = coerceAgainstSchema(input, alt, root, depth + 1, refStack);
+        if (result !== input) return result;
+      }
```
[the remark this answers is not in the document]
Comment (artiom) [reply src/shared/utils/schema-validator.ts:268]: Fixed in eddf8a54 (`schema-validator.ts:195-211`). Both Array and object branches now return the result of the FIRST matching-type alternative regardless of whether anything was coerced — matching the documented "declared order wins" contract. ajv still enforces the actual schema match at validate time, so an unsuitable first alt still produces a 400.

**Comment by artiom** [comment]
Re: CodeRabbit's duplicate finding on `schema-validator.ts:195-208` — the round-3 fix ("first matching-type wins") regressed the `anyOf: [{required:[a]}, {required:[b]}]` case where input only matches the later alt.

Fixed in a6ee03c5 with a sharper rule: when picking an object alternative, skip alts whose `required` keys are missing from the input. This now correctly:

- `{b:'1'}` vs `[{required:[a]}, {required:[b]}]` → picks alt B, coerces `b` to number ✓
- `{id:'1'}` vs `[{id:string}, {id:number}]` → picks alt A (no required), `id` stays string ✓

New regression test added ("skips anyOf alternatives whose required keys are missing"). All 22 validation tests green, generated artifacts byte-identical to main.

File: src/shared/utils/schema-validator.ts [hunk 7]
```diff
@@ -0,0 +1,382 @@
...
+    if (!s) return true;
+  }
+  const required = s.required;
+  if (Array.isArray(required)) {
+    for (const key of required) {
+      if (typeof key === 'string' && !(key in input)) return false;
+    }
+  }
+  return true;
+};
```
[the remark this answers is not in the document]
Comment (artiom) [reply src/shared/utils/schema-validator.ts:142]: Fixed in 7d2e820b using your suggested patch. `collectRequiredKeys` now walks `$ref` chains and `allOf` recursively. Also extended `inferExpectedType` to descend through `allOf` so alts whose type comes from a subschema are correctly identified as object/array. Added regression test "resolves required keys transitively through $ref and allOf for anyOf skip logic".

**Review by andyMrtnzP — APPROVED** [review]
